cmd/rook/ceph/osd.go

```diff
@@ -418,35 +416,54 @@ func readCephSecret(path string) error {
 	return nil
 }
 
-func destroyOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) (*oposd.OSDInfo, error) {
-	osdInfo, err := osddaemon.GetOSDInfoById(context, clusterInfo, osdID)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get OSD info for OSD.%d", osdID)
+// destroyOSD fetches the OSD to be replaced based on the ID and then destroys that OSD and zaps the backing disk
+func destroyOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, id int) (*osd.OSDReplaceInfo, error) {
+	var osd *osd.OSDReplaceInfo
+	var err error
+	if cfg.pvcBacked {
+		osd = &oposd.OSDReplaceInfo{ID: id, Path: fmt.Sprintf("/mnt/%s", os.Getenv(oposd.PVCNameEnvVarName))}
+	} else {
+		osd, err = osddaemon.GetOSDReplaceInfoById(context, clusterInfo, id)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get OSD info for OSD.%d", id)
+		}
 	}
 
-	// destroy the osd
-	logger.Infof("destroying OSD %d on path %q in %q mode", osdInfo.ID, osdInfo.BlockPath, osdInfo.CVMode)
-	destroyOSDArgs := []string{"osd", "destroy", fmt.Sprintf("osd.%d", osdInfo.ID), "--yes-i-really-mean-it"}
+	logger.Infof("destroying OSD %d", osd.ID)
+	destroyOSDArgs := []string{"osd", "destroy", fmt.Sprintf("osd.%d", osd.ID), "--yes-i-really-mean-it"}
 	_, err = client.NewCephCommand(context, clusterInfo, destroyOSDArgs).Run()
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to destroy osd.%d.", osdInfo.ID)
+		return nil, errors.Wrapf(err, "failed to destroy osd.%d.", osd.ID)
 	}
+	logger.Infof("successfully destroyed osd.%d", osd.ID)
 
-	// Sanitize OSD disk
-	s := cleanup.NewDiskSanitizer(context, clusterInfo,
-		&cephv1.SanitizeDisksSpec{
-			Method:     cephv1.SanitizeMethodProperty(cephv1.SanitizeMethodComplete),
-			DataSource: cephv1.SanitizeDataSourceProperty(cephv1.SanitizeDataSourceZero),
-			Iteration:  1,
-		},
-	)
+	// remove the dm device
+	if cfg.pvcBacked && cfg.storeConfig.EncryptedDevice {
+		target := oposd.EncryptionDMName(os.Getenv(oposd.PVCNameEnvVarName), oposd.DmcryptBlockType)
+		err = osddaemon.RemoveEncryptedDevice(context, target)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to remove dm device %q", target)
+		}
+	}
 
-	// TODO: handle disk sanitization errors
-	if osdInfo.CVMode == "raw" {
-		s.SanitizeRawDisk([]oposd.OSDInfo{osdInfo})
-	} else if osdInfo.CVMode == "lvm" {
-		s.SanitizeLVMDisk([]oposd.OSDInfo{osdInfo})
+	// zap the device
+	output, err := context.Executor.ExecuteCommandWithCombinedOutput("stdbuf", "-oL", "ceph-volume", "lvm", "zap", osd.Path, "--destroy")
```

Comment:
To zap and wipe the disk, can we call helpers in `pkg/daemon/ceph/cleanup/`? It's nice to keep the `cmd` packages only focusing on command line processing.

Reply:
`pkg/daemon/ceph/cleanup/` is used by the cleanup jobs that run at the end so it just logs the error. It does not return any error.  We need error handling while zapping a single disk.